### PR TITLE
[Enhancement] Adaptive tablet compaction concurrency for cloud native table

### PR DESCRIPTION
**Pull request by sduzh**
If the concurrency of the compaction is too high, the compaction tasks may
fail due to memory limit exceeded and if the concurrency is too low, it
cannot fully utilize resources. Therefore, a simple adaptive strategy is added:
- The initial maximum compaction concurrency is config::compact_threads.
- Once Status::MemoryLimitExceeded is encountered, reduce the maximum
   concurrency by one until the concurrency is reduced to 1.
- If there is no Status::MemoryLimitExceeded encountered for a period of
  time, gradually increase the maximum concurrency until
  config::compact_threads is reached or Status::MemoryLimitExceeded is
  encountered again.

## What type of PR is this：
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto backported to target branch
  - [x] 3.0
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

File: be/src/storage/lake/compaction_scheduler.cpp [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+        context->start_time.store(start_time, std::memory_order_relaxed);
+    }
+    const auto runs = context->runs.fetch_add(1, std::memory_order_relaxed);
+
+    auto status = Status::OK();
+    if (config::lake_compaction_check_txn_log_first && runs == 0 && txn_log_exists(tablet_id, txn_id)) {
+        context->skipped.store(true, std::memory_order_relaxed);
+        context->progress.update(100);
+        VLOG(2) << "Skipped already succeeded compaction task. tablet_id=" << tablet_id << " txn_id=" << txn_id
+                << " _version=" << version;
```
Comment (wyb) [inline be/src/storage/lake/compaction_scheduler.cpp:199]:
```suggestion
                << " version=" << version;
```

File: be/src/storage/lake/compaction_scheduler.h [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+    // with bthread::Mutex and replace std::condition_variable with bthread::ConditionVariable:
+    // ```
+    // using TaskQueue = BlockingQueue<CompactionStatePtr,
+    //                                 std::deque<CompactionStatePtr>,
+    //                                 bthread::Mutex,
+    //                                 bthread::ConditionVariable>;
+    // ```
+    using TaskQueue = UnboundedBlockingQueue<CompactionContextPtr>;
+
+    void remove_states(const std::vector<std::unique_ptr<CompactionTaskContext>>& contexes);
```
Comment (wyb) [inline be/src/storage/lake/compaction_scheduler.h:174]:
```suggestion
    void remove_states(const std::vector<std::unique_ptr<CompactionTaskContext>>& contexts);
```

File: be/src/storage/lake/compaction_scheduler.h [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+    void steal_task(int start_index, std::unique_ptr<CompactionTaskContext>* context);
+
+    int choose_task_queue_by_txn_id(int64_t txn_id) const { return txn_id % _task_queue_count; }
+
+    bool txn_log_exists(int64_t tablet_id, int64_t txn_id) const;
+
+    TabletManager* _tablet_mgr;
+    Limiter _limiter;
+    bthread::Mutex _states_lock;
+    butil::LinkedList<CompactionTaskContext> _contexs;
```
Comment (wyb) [inline be/src/storage/lake/compaction_scheduler.h:189]:
```suggestion
    butil::LinkedList<CompactionTaskContext> _contexts;
```

File: be/src/storage/lake/compaction_scheduler.cpp (hunk 1 above)
Comment (sduzh) [reply be/src/storage/lake/compaction_scheduler.cpp]: Fixed

File: be/src/storage/lake/compaction_scheduler.h (hunk 2 above)
Comment (sduzh) [reply be/src/storage/lake/compaction_scheduler.h:174]: Fixed

File: be/src/storage/lake/compaction_scheduler.h (hunk 3 above)
Comment (sduzh) [reply be/src/storage/lake/compaction_scheduler.h]: Fxied

**Review by wyb — APPROVED** [review]

**Review by gengjun-git — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.0

**Pull request merged**